File: src/main/kotlin/org/yc7521/pay/util/ResponseUtil.java
package org.yc7521.pay.util;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.ResponseEntity;

import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.List;
import java.util.Map;

public class ResponseUtil {
  public static void write(ResponseEntity<?> responseEntity,
                           HttpServletResponse servletResponse) throws IOException {
    for (Map.Entry<String, List<String>> header : responseEntity
      .getHeaders()
      .entrySet()) {
      String headerKey = header.getKey();
      for (String value : header.getValue()) {
        servletResponse.addHeader(headerKey, value);
      }
    }

    servletResponse.setStatus(responseEntity.getStatusCodeValue());
    final String body = new ObjectMapper().writeValueAsString(responseEntity.getBody());
    if (body != null) {
      servletResponse.getWriter().write(body);
    }
  }
}

File: src/main/kotlin/org/yc7521/pay/util/ResponseUtil.kt
package org.yc7521.pay.util

import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.http.ResponseEntity
import java.io.OutputStreamWriter
import javax.servlet.http.HttpServletResponse

object ResponseUtil {
  @JvmStatic
  fun write(
    responseEntity: ResponseEntity<*>,
    servletResponse: HttpServletResponse,
  ) {
    for ((headerKey, value1) in responseEntity.headers) {
      for (value in value1) {
        servletResponse.addHeader(headerKey, value)
      }
    }
    servletResponse.status = responseEntity.statusCodeValue
    ObjectMapper().writeValueAsString(responseEntity.body)?.let { body ->
      OutputStreamWriter(
        servletResponse.outputStream, Charsets.UTF_8
      ).use { it.write(body) }
    }
    servletResponse.characterEncoding = "UTF-8"
  }
}
